### PR TITLE
Cache final chain data

### PR DESCRIPTION
**Pull request by kstdl**
#1896

### CODE REVIEW

File: libraries/core_libs/consensus/include/final_chain/final_chain.hpp [hunk 1]
```diff
@@ -65,7 +65,7 @@ class FinalChain {
    * @param n block number of header to get. If not specified then it returns latest
    * @return BlockHeader
    */
-  virtual std::shared_ptr<BlockHeader const> block_header(std::optional<EthBlockNumber> n = {}) const = 0;
+  virtual std::shared_ptr<BlockHeader> block_header(std::optional<EthBlockNumber> n = {}) const = 0;
```
[the remark this answers is not in the document]
Comment (kstdl) [reply libraries/core_libs/consensus/include/final_chain/final_chain.hpp]: Added const to caches where it makes sense

File: libraries/core_libs/consensus/include/final_chain/cache.hpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  MapByBlockCache(const MapByBlockCache &) = delete;
+  MapByBlockCache(MapByBlockCache &&) = delete;
+  MapByBlockCache &operator=(const MapByBlockCache &) = delete;
+  MapByBlockCache &operator=(MapByBlockCache &&) = delete;
+
+  MapByBlockCache(uint64_t blocks_to_save, GetterFn getter_fn) : kBlocksToKeep(blocks_to_save), getter_fn_(getter_fn) {}
+
+  void cleanup() {
+    if (data_by_block_.size() > kBlocksToKeep) {
+      data_by_block_.erase(data_by_block_.begin());
```
[the remark this answers is not in the document]
Comment (kstdl) [reply libraries/core_libs/consensus/include/final_chain/cache.hpp]: This method is unused, removed it

[the remark this answers is not in the document]
Comment (kstdl) [reply libraries/core_libs/consensus/include/final_chain/cache.hpp]: But it is protected already. Or am I missing something?

File: libraries/core_libs/consensus/include/final_chain/cache.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  using GetterFn = std::function<Value(uint64_t, const Key &)>;
+  using ValueMap = std::unordered_map<Key, Value>;
+  using DataMap = std::map<uint64_t, ValueMap>;
+
+  MapByBlockCache(const MapByBlockCache &) = delete;
+  MapByBlockCache(MapByBlockCache &&) = delete;
+  MapByBlockCache &operator=(const MapByBlockCache &) = delete;
+  MapByBlockCache &operator=(MapByBlockCache &&) = delete;
+
+  MapByBlockCache(uint64_t blocks_to_save, GetterFn getter_fn)
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/include/final_chain/cache.hpp:42]: GetterFn&& ?

File: libraries/core_libs/consensus/src/final_chain/final_chain.cpp [hunk 4]
```diff
@@ -32,62 +30,101 @@ class FinalChainImpl final : public FinalChain {
...
+        accounts_cache_(config.final_chain_cache_in_blocks,
+                        [this](uint64_t blk, const addr_t& addr) { return state_api_.get_account(blk, addr); }),
+
+        total_vote_count_cache_(config.final_chain_cache_in_blocks,
+                                [this](uint64_t blk) { return get_dpos_eligible_total_vote_count(blk); }),
+        dpos_vote_count_cache_(
+            config.final_chain_cache_in_blocks,
+            [this](uint64_t blk, const addr_t& addr) { return get_dpos_eligible_vote_count(blk, addr); }),
+        dpos_is_eligible_cache_(config.final_chain_cache_in_blocks,
+                                [this](uint64_t blk, const addr_t& addr) { return get_dpos_is_eligible(blk, addr); }) {
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/final_chain/final_chain.cpp:75]: what do we need this get_dpos_is_elogible function ? cannot we use here directly return state_api_.dpos_is_eligible(blk_num, addr) ?

**Review by JakubFornadel — CHANGES_REQUESTED** [review]
I dont see the changes related to deleting the current "cache" we are using. After this proper cache is used, we should be able to completely get rid of:

- size_t getDposTotalVotesCount() const;
- size_t getDposWeightedVotesCount() const
- uint64_t getCurrentPbftSortitionThreshold(PbftVoteTypes vote_type) const;
- potentially also other functions and variables they work with

so we celan the code little bit and we use only getters with specified period ?

Also pls look for TODO[1896] and remove it

File: libraries/core_libs/consensus/include/final_chain/cache.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kstdl) [reply libraries/core_libs/consensus/include/final_chain/cache.hpp]: lock was wrongly placed in `MapByBlockCache::get`. Should protect also call of find. Fixed

**Pull request merged**